firedrake/slate/slac/utils.py

```diff
@@ -133,3 +135,43 @@ def visit_Symbol(self, o, *args, **kwargs):
             return o
 
         return SymbolWithFuncallIndexing(o.symbol, o.rank, o.offset)
+
+
+# Thanks, Miklos!
+def traverse_dags(expr_dags):
+    """Traverses a DAG and returns the unique operands associated
+    with the DAG.
+
+    :arg expr_dags: an iterable of Slate nodes that make up the
+                    DAG of a given Slate expression.
+    """
+    seen = set()
+    container = []
+
+    for tensor in expr_dags:
+        if tensor not in seen:
+            seen.add(tensor)
+            container.append(tensor)
+
+    while container:
+        tensor = container.pop()
+        yield tensor
+        for operand in tensor.operands:
+            if operand not in seen:
+                seen.add(operand)
+                container.append(operand)
+
+
+# This function is based on the GEM DAG reference count
+# algorithm in TSFC and adapted for Slate tensors.
+def collect_reference_count(expr_dags):
+    """Returns a mapping from operand to reference count.
+
+    :arg expr_dags: an iterable of Slate nodes that make up the
+                    DAG of a given Slate expression.
+    """
+    result = collections.Counter(expr_dags)
+    for tensor in traverse_dags(expr_dags):
+        result.update(tensor.operands)
+
+    return result
```

Comment:
It is probably beyond the scope of this pull request, so I'm just telling for later consideration: if the slate root class (`TensorBase`) were derived from `gem.node.Node` (not to be confused with `gem.gem.Node`) and followed the same conventions, then you could just use the corresponding utility functions from `gem.node` rather than having to implement them again.

I suggested this at the beginning, but it was dismissed then for the difficulty of inheriting both `ufl.form.Form` and `gem.node.Node`. However, slate tensors have not been derived classes of `ufl.form.Form` for quite long, so this objection does not exist anymore.